### DISCUSSION TITLE
Using the envelope parameters function in PyAT

### DISCUSSION THREAD

**Discussion by elafmusa** [atcollab/at#293] [Python AT]
Hi everyone,

I am tiring to get the ring parameters using the envelope parameters function in PyAT, i run the following code first with setting the radiation off but i got an error massage that it should be on, so i turned it on and the error now is " singular matrix" Could you please give feedback on this.
Thank you in advance

_**from at import Drift, Quadrupole, Bend, Corrector, Sextupole, Monitor

E0 = 3e9 
Ncells = 30
Kq = 0.1*np.sqrt(2) 
L = 3
ksf = 0.00
ksd = -0.00

Dr = Drift('Dr', L/8)
Drh = Drift('Dr', Dr.Length/2)
QF1 = Quadrupole('QF1', L/4, Kq,
PassMethod = 'StrMPoleSymplectic4Pass')
QD2 = Quadrupole('QD2', L/4, -Kq,
PassMethod = 'StrMPoleSymplectic4Pass')
QD3 = Quadrupole('QD3', L/4, -Kq,
PassMethod = 'StrMPoleSymplectic4Pass')
QF4 = Quadrupole('QF4', L/4, Kq,
PassMethod = 'StrMPoleSymplectic4Pass')
Dip = Bend('Bend', L/2, 2*np.pi/(2*Ncells), 
PassMethod = 'BndMPoleSymplectic4Pass')
SF = Sextupole('SF', Dr.Length/2, ksf,
PassMethod = 'StrMPoleSymplectic4Pass')
SD = Sextupole('SD', Dr.Length/2, ksd,
PassMethod = 'StrMPoleSymplectic4Pass')

BPM1 = Monitor('BPM1')
BPM2 = Monitor('BPM2')
BPM3 = Monitor('BPM3')
BPM4 = Monitor('BPM4')

from at import RFCavity
Lc = 0 #Cavity Length
Vc = 1e6 #Cavity Voltage
fc = ring.circumference/299792458 #Cavity Frequency
h = 1 #Harmonic Number
#rfc = RFCavity('CA', Lc, Vc, fc, h, E0, PassMethod = 'CavityPass')
rfc = RFCavity('CA', Lc, Vc, fc, h, E0, PassMethod = 'CavityPass')

arc = [rfc,Dr, Dr, Dr, BPM1, Dr, Dr, Dr, QF1, Dr, Dr, QD2, Dr, Dr, Dip, BPM2, Dip, Drh, QD3, Drh, SD, Drh, Drh, Drh, BPM3, QF4, Drh, SF]
arc = at.Lattice(arc, energy=E0, periodicity = Ncells)
ring = at.Lattice(arc * int(Ncells/2),energy=E0, periodicity=1)
ring = at.Lattice(ring)
#ring.radiation_off()

ring.radiation_on()

env = ring.envelope_parameters()](url)**_

**Comment by lfarv** [comment]
I made a few modifications to your example:
- added missing imports,
- removed useless default values for element creation,
- set the cavity "inactive" first, to be in the state "radiation OFF"
- used set_cavity to set the cavity frequency

and I get it running:


```python
import at
from at import Drift, Quadrupole, Bend, Corrector, Sextupole, Monitor, RFCavity
import numpy as np
```


```python
E0 = 3e9
Ncells = 30
Kq = 0.1*np.sqrt(2)
L = 3
ksf = 0.00
ksd = -0.00

Dr = Drift('Dr', L/8)
Drh = Drift('Dr', Dr.Length/2)
QF1 = Quadrupole('QF1', L/4, Kq)
QD2 = Quadrupole('QD2', L/4, -Kq)
QD3 = Quadrupole('QD3', L/4, -Kq)
QF4 = Quadrupole('QF4', L/4, Kq)
Dip = Bend('Bend', L/2, 2*np.pi/(2*Ncells))
SF = Sextupole('SF', Dr.Length/2, ksf)
SD = Sextupole('SD', Dr.Length/2, ksd)
BPM1 = Monitor('BPM1')
BPM2 = Monitor('BPM2')
BPM3 = Monitor('BPM3')
BPM4 = Monitor('BPM4')
```

```python
Lc = 0 #Cavity Length
Vc = 1e6 #Cavity Voltage
fc = 0
h = 1 #Harmonic Number
#rfc = RFCavity('CA', Lc, Vc, fc, h, E0, PassMethod = 'CavityPass')
rfc = RFCavity('CA', Lc, Vc, fc, h, E0, PassMethod = 'IdentityPass')

arc = [rfc,Dr, Dr, Dr, BPM1, Dr, Dr, Dr, QF1, Dr, Dr, QD2, Dr, Dr, Dip, BPM2, Dip, Drh, QD3, Drh, SD, Drh, Drh, Drh, BPM3, QF4, Drh, SF]
arc = at.Lattice(arc, energy=E0, periodicity = Ncells)
ring = at.Lattice(arc * int(Ncells/2),energy=E0, periodicity=1)
ring.set_cavity(Frequency=at.Frf.NOMINAL)
```


```python
print(arc.radiation_parameters())
```

                  Frac. tunes: [0.36108352 0.78177024 0.05995418]
                        Tunes: [3.36108352 2.78177024]
               Chromaticities: [-2.87866976 -2.80719782]
     Momentum compact. factor: 7.529361e-02
                  Slip factor: -7.529358e-02
                       Energy: 3.000000e+09 eV
           Energy loss / turn: 5.002450e+05 eV
     Radiation integrals - I1: 25.411593120922145 m
                           I2: 0.43864908449286033 m^-1
                           I3: 0.030623483141036856 m^-2
                           I4: 0.12385302286663964 m^-1
                           I5: 0.03687035913499427 m^-1
              Mode emittances: [1.54692449e-06            nan            nan]
    Damping partition numbers: [0.71764897 1.         2.28235103]
                Damping times: [0.01881523 0.01350273 0.00591615] s
                Energy spread: 0.000635606
                 Bunch length: 0.0428766 m
             Cavities voltage: 30000000.0 V
            Synchrotron phase: 3.12492 rd
        Synchrotron frequency; 53255.7 Hz



```python
ring.radiation_on()
print(ring.envelope_parameters())
```

      Frac. tunes (6D motion): [0.68052375 0.39088639 0.00774013]
                       Energy: 3.000000e+09 eV
           Energy loss / turn: 2.501225e+05 eV
              Mode emittances: [1.54706394e-06 1.45853938e-35 1.05549335e-04]
    Damping partition numbers: [0.71761354 0.99999978 2.28238668]
                Damping times: [0.0188166  0.01350305 0.0059162 ] s
                Energy spread: 0.000635624
                 Bunch length: 0.166073 m
             Cavities voltage: 15000000.0 V
            Synchrotron phase: 3.12492 rd
        Synchrotron frequency; 13750.7 Hz

**Comment by elafmusa** [comment]
@lfarv Thank you very much, i still have a problem when i run the code, an error of " 'Lattice' object has no attribute 'set_cavity'" arises from the line :
ring.set_cavity(Frequency=at.Frf.NOMINAL)
